cmd/auth-handler.go

```diff
@@ -174,13 +196,26 @@ func checkRequestAuthType(ctx context.Context, r *http.Request, action policy.Ac
 		r.Body = ioutil.NopCloser(bytes.NewReader(payload))
 	}
 
+	var claims map[string]interface{}
+	if sessionToken := r.Header.Get("X-Amz-Security-Token"); sessionToken != "" {
+		data, err := base64.URLEncoding.DecodeString(r.Header.Get("X-Amz-Security-Token"))
+		if err != nil {
+			logger.LogIf(context.Background(), err)
+			return ErrAccessDenied
+		}
+		if err = json.Unmarshal(data, &claims); err != nil {
+			logger.LogIf(context.Background(), err)
+			return ErrAccessDenied
+		}
+	}
 	if globalPolicySys.IsAllowed(policy.Args{
 		AccountName:     accountName,
 		Action:          action,
 		BucketName:      bucketName,
 		ConditionValues: getConditionValues(r, locationConstraint),
-		IsOwner:         isOwner,
+		IsOwner:         accountName != "",
 		ObjectName:      objectName,
+		Claims:          claims,
```

Comment:
Claims should be protected by a signature, we can't trust the client sent claims.

Reply:
done